File: examples/ecommerce-product-search/typescript/src/ingest.ts
import fs from "fs";
import path from "path";
import yaml from "js-yaml";
import { SIEClient } from "@superlinked/sie-sdk";
import type { Entity } from "@superlinked/sie-sdk";

import { fileURLToPath } from "url";

// Config matches config.yaml structure exactly
interface Config {
  cluster: {
    url: string;
    api_key: string;
    gpu: string;
    provision_timeout_s: number;
  };
  models: {
    embedding: string;
    reranker: string;
    extractor: string;
  };
  extract_labels: string[];
  search: {
    top_k_candidates: number;
    top_k_results: number;
  };
  ingest: {
    batch_size_extraction: number;
    batch_size_encoding: number;
    confidence_threshold: number;
    junk_text_max_len?: number;
  };
}

interface Product {
  id: string;
  title: string;
  description: string;
  category: string;
  price: number;
  rating: number;
  features?: string[];
  image?: string;
}

interface ExtractedProduct extends Product {
  attributes: Record<string, string>;
}

// Resolve paths relative to this file's location (works regardless of cwd)
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, "..", "..");
const TS_DIR = path.resolve(__dirname, "..");

function loadConfig(): Config {
  const data = fs.readFileSync(path.join(ROOT, "config.yaml"), "utf-8");
  return yaml.load(data) as Config;
}

function loadProducts(): Product[] {
  const data = fs.readFileSync(path.join(ROOT, "data", "products.json"), "utf-8");
  return JSON.parse(data);
}

function buildAttributes(
  entities: Entity[],
  labels: string[],
  threshold: number,
  junkLen: number
): Record<string, string> {
  const best: Record<string, { text: string; score: number }> = {};

  for (const e of entities) {
    const text = e.text.trim();
    if (e.score < threshold) continue;
    if (text.length > junkLen && !text.includes(" ")) continue;
    if (!labels.includes(e.label)) continue;
    if (!(e.label in best) || e.score > best[e.label].score) {
      best[e.label] = { text, score: e.score };
    }
  }

  return Object.fromEntries(Object.entries(best).map(([k, v]) => [k, v.text]));
}

async function extractAttributes(
  client: SIEClient,
  products: Product[],
  config: Config
): Promise<Record<string, Record<string, string>>> {
  const model = config.models.extractor;
  const labels = config.extract_labels;
  const batchSize = config.ingest.batch_size_extraction;
  const threshold = config.ingest.confidence_threshold;
  const junkLen = config.ingest.junk_text_max_len ?? 15;
  const total = products.length;

  console.log(`\n--- Extraction (${model}) ---`);
  console.log(`  Labels: ${labels.join(", ")}`);

  const extracted: Record<string, Record<string, string>> = {};

  for (let i = 0; i < total; i += batchSize) {
    const batch = products.slice(i, Math.min(i + batchSize, total));
    const items = batch.map((p) => ({ id: p.id, text: p.description.substring(0, 512) }));
    const t0 = Date.now();

    const results = await client.extract(model, items, { labels });

    const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
    console.log(`  [${Math.min(i + batchSize, total)}/${total}] batch in ${elapsed}s`);

    results.forEach((result, idx) => {
      const productId = batch[idx].id;
      extracted[productId] = buildAttributes(result.entities, labels, threshold, junkLen);
    });
  }

  return extracted;
}

async function encodeProducts(
  client: SIEClient,
  products: Product[],
  config: Config
): Promise<number[][]> {
  const model = config.models.embedding;
  const batchSize = config.ingest.batch_size_encoding;
  const total = products.length;

  console.log(`\n--- Encoding (${model}) ---`);

  const allEmbeddings: number[][] = [];

  for (let i = 0; i < total; i += batchSize) {
    const batch = products.slice(i, Math.min(i + batchSize, total));
    const items = batch.map((p) => ({
      id: p.id,
      text: `${p.title}. ${p.description.substring(0, 512)}`,
    }));
    const t0 = Date.now();

    const results = await client.encode(model, items);

    const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
    console.log(`  [${Math.min(i + batchSize, total)}/${total}] batch in ${elapsed}s`);

    // Fail loud on missing dense vectors. Silently skipping here would desync
    // embeddings from metadata (which is built positionally from products).
    for (const r of results) {
      if (!r.dense) {
        throw new Error(`encode() returned no dense vector for item ${r.id ?? "<unknown>"}`);
      }
      allEmbeddings.push(Array.from(r.dense));
    }
  }

  return allEmbeddings;
}

async function main(): Promise<void> {
  try {
    console.log("=== E-Commerce Product Search (TypeScript Ingest) ===\n");

    const config = loadConfig();
    const clusterUrl = process.env.SIE_CLUSTER_URL ?? config.cluster.url;
    const apiKey = process.env.SIE_API_KEY ?? config.cluster.api_key;

    console.log(`Cluster: ${clusterUrl}`);
    console.log(`Embedder: ${config.models.embedding}`);
    console.log(`Extractor: ${config.models.extractor}`);

    const gpu = config.cluster.gpu;
    const provisionTimeout = (config.cluster.provision_timeout_s ?? 600) * 1000;
    const client = new SIEClient(clusterUrl, { apiKey, gpu, waitForCapacity: true, provisionTimeout, timeout: 300_000 });

    console.log("\nLoading products...");
    const products = loadProducts();
    console.log(`Loaded ${products.length} products`);

    // Phase 1: Extract attributes
    const extracted = await extractAttributes(client, products, config);
    console.log(`\nExtraction complete`);

    // Phase 2: Encode embeddings
    const embeddings = await encodeProducts(client, products, config);
    console.log(`\nEncoding complete: ${embeddings.length} embeddings, ${embeddings[0]?.length}d`);

    // Build metadata with attributes
    const metadata: ExtractedProduct[] = products.map((p) => ({
      ...p,
      description: p.description.substring(0, 500),
      attributes: extracted[p.id] ?? {},
    }));

    // Save output to data-ts/ inside typescript/
    const outDir = path.join(TS_DIR, "data-ts");
    if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });

    fs.writeFileSync(path.join(outDir, "embeddings.json"), JSON.stringify(embeddings));
    console.log(`\nSaved embeddings → data-ts/embeddings.json`);

    fs.writeFileSync(path.join(outDir, "metadata.json"), JSON.stringify(metadata, null, 2));
    console.log(`Saved metadata   → data-ts/metadata.json (${metadata.length} products)`);

    console.log("\n=== Ingest Complete ===");
  } catch (error) {
    console.error("Fatal error:", error);
    process.exit(1);
  }
}

main();

File: examples/ecommerce-product-search/python/ingest.py
"""Ingest pipeline: extract attributes + encode embeddings for all products."""

import json
import os
import time
from pathlib import Path

import numpy as np
import yaml

from sie_sdk import SIEClient
from sie_sdk.types import Item


def load_config():
    root = Path(__file__).resolve().parent.parent
    return yaml.safe_load((root / "config.yaml").read_text())


def load_products():
    path = Path(__file__).resolve().parent.parent / "data" / "products.json"
    with open(path) as f:
        return json.load(f)


def batch_iter(items, batch_size):
    for i in range(0, len(items), batch_size):
        yield items[i : i + batch_size]


def run_extraction(client, config, products):
    """Run extract() on all products to get structured attributes."""
    model = config["models"]["extractor"]
    labels = config["extract_labels"]
    gpu = config["cluster"]["gpu"]
    timeout = config["cluster"]["provision_timeout_s"]

    print(f"\n--- Extraction ({model}) ---")
    print(f"  Labels: {labels}")

    extracted = {}
    batch_size = config.get("ingest", {}).get("batch_size_extraction", 8)
    threshold = config.get("ingest", {}).get("confidence_threshold", 0.5)
    junk_len = config.get("ingest", {}).get("junk_text_max_len", 15)
    total = len(products)

    for i, batch in enumerate(batch_iter(products, batch_size)):
        items = [Item(id=p["id"], text=p["description"][:512]) for p in batch]
        start = time.time()

        results = client.extract(
            model,
            items,
            labels=labels,
            gpu=gpu,
            wait_for_capacity=True,
            provision_timeout_s=timeout,
        )

        elapsed = time.time() - start
        done = min((i + 1) * batch_size, total)
        print(f"  [{done}/{total}] batch in {elapsed:.1f}s")

        for result in results:
            item_id = result.get("id", "")
            entities = result.get("entities", [])
            attrs = {}
            for e in entities:
                label = e["label"]
                text = e["text"].strip()
                if e["score"] < threshold:
                    continue
                if len(text) > junk_len and " " not in text:
                    continue
                if label not in attrs or e["score"] > attrs[label]["score"]:
                    attrs[label] = {"text": text, "score": e["score"]}
            extracted[item_id] = {k: v["text"] for k, v in attrs.items()}

    return extracted


def run_encoding(client, config, products):
    """Run encode() on all products to get dense embeddings."""
    model = config["models"]["embedding"]
    gpu = config["cluster"]["gpu"]
    timeout = config["cluster"]["provision_timeout_s"]

    print(f"\n--- Encoding ({model}) ---")

    embeddings = []
    batch_size = config.get("ingest", {}).get("batch_size_encoding", 32)
    total = len(products)

    for i, batch in enumerate(batch_iter(products, batch_size)):
        items = [Item(id=p["id"], text=f"{p['title']}. {p['description'][:512]}") for p in batch]
        start = time.time()

        results = client.encode(
            model,
            items,
            gpu=gpu,
            wait_for_capacity=True,
            provision_timeout_s=timeout,
        )

        elapsed = time.time() - start
        done = min((i + 1) * batch_size, total)
        print(f"  [{done}/{total}] batch in {elapsed:.1f}s")

        # The server preserves input order (encode.py zips results to input by index),
        # so results[j] corresponds to batch[j], and the final embeddings array is
        # aligned positionally with the products list.
        if len(results) != len(batch):
            raise ValueError(
                f"encode() returned {len(results)} results for a batch of {len(batch)}"
            )
        for result in results:
            if "dense" not in result:
                raise ValueError(f"encode() returned no dense vector for item {result.get('id', '<unknown>')}")
            embeddings.append(result["dense"])

    if not embeddings:
        raise ValueError("No embeddings were generated; check products list and cluster connection")

    return np.stack(embeddings)


def main():
    config = load_config()
    products = load_products()
    print(f"Loaded {len(products)} products")

    cluster_url = os.environ.get("SIE_CLUSTER_URL", config["cluster"]["url"])
    api_key = os.environ.get("SIE_API_KEY", config["cluster"]["api_key"])

    with SIEClient(cluster_url, api_key=api_key) as client:
        # Step 1: Extract structured attributes
        extracted_attrs = run_extraction(client, config, products)

        # Step 2: Encode embeddings
        embeddings = run_encoding(client, config, products)
        print(f"\n  Embeddings shape: {embeddings.shape}")

    # Step 3: Build metadata (product info + extracted attributes)
    metadata = []
    for p in products:
        entry = {
            "id": p["id"],
            "title": p["title"],
            "description": p["description"][:500],
            "category": p["category"],
            "price": p["price"],
            "rating": p["rating"],
            "image": p["image"],
            "features": p.get("features", []),
            "attributes": extracted_attrs.get(p["id"], {}),
        }
        metadata.append(entry)

    # Save
    data_dir = Path(__file__).resolve().parent.parent / "data"
    np.save(data_dir / "embeddings.npy", embeddings)
    with open(data_dir / "metadata.json", "w") as f:
        json.dump(metadata, f, indent=2)

    print(f"\n  Saved embeddings to data/embeddings.npy")
    print(f"  Saved metadata to data/metadata.json")

    # Print a few examples
    print("\n--- Sample extractions ---")
    for p in products[:5]:
        attrs = extracted_attrs.get(p["id"], {})
        if attrs:
            print(f"  {p['title'][:60]}")
            print(f"    -> {attrs}")


if __name__ == "__main__":
    main()
